docs/tutorials/shors-algorithm.ipynb

```diff
@@ -0,0 +1,1334 @@
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "<Image src=\"/docs/images/tutorials/shors-algorithm/extracted-outputs/559d7030-1f67-44e8-afa7-6afc7a334677-0.avif\" alt=\"Output of the previous code cell\" />"
+      ]
+     },
+     "execution_count": 26,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "plot_histogram(counts, figsize=(35, 5))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3614aa7e-0410-4e9e-b33a-3caf407e086a",
+   "metadata": {},
+   "source": [
+    "As we can see, we obtained the same bitstrings with highest counts. Since quantum hardware has noise, there is some leakage to other bitstrings, which we can filter out statistically."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "id": "f9cf9a07-5251-47bc-9713-c802f8f1a37c",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{'00000000': 58, '01000000': 41, '11000000': 42, '10000000': 40}\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Dictionary of bitstrings and their counts to keep\n",
+    "counts_keep = {}\n",
+    "# Threshold to filter\n",
+    "threshold = np.max(list(counts.values())) / 2\n",
+    "\n",
+    "for key, value in counts.items():\n",
+    "    if value > threshold:\n",
+    "        counts_keep[key] = value\n",
+    "\n",
+    "print(counts_keep)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c32cca67-9c9d-4373-9a0a-3174e884dd91",
+   "metadata": {},
+   "source": [
+    "## Step 4: Post-process and return result in desired classical format"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5a7b116b-02d0-47dc-8cab-e792ba36d868",
+   "metadata": {},
+   "source": [
+    "### Integer Factorization"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "691df180-6ac3-49ac-ad4d-2d78348ceea7",
+   "metadata": {},
+   "source": [
+    "So far, we discussed how we can implement the order finding problem using a phase estimation circuit. Now, we connect the order finding problem to integer factorization, which completes Shor's algorithm. Note that this part of the algorithm is classical."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fe3472a3-f8cd-4db9-8178-c025ccf2ec55",
+   "metadata": {},
+   "source": [
+    "We now demonstrate this using our example of $N = 15$ and $a = 2$. Recall that the phase we measured is $k / r$ where $$a^r \\; (\\textrm{mod} \\; N) = 1$$ and $k$ is a random integer between $0$ and $r - 1$. From this equation, we have $$(a^r - 1) \\; (\\textrm{mod} \\; N) = 0,$$ which means $N$ must divide $a^r-1$. And if $r$ is also even, then we can write $$a^r -1 = (a^{r/2}-1)(a^{r/2}+1).$$ If $r$ is not even, we cannot go further and must try again with a different value for $a$. Otherwise there is a high probability that the greatest common divisor of $N$ and either $a^{r/2}-1$, or $a^{r/2}+1$ is a proper factor of $N$.\n",
```

Comment:
```suggestion
    "We now demonstrate this using our example of $N = 15$ and $a = 2$. Recall that the phase we measured is $k / r$, where $$a^r \\; (\\textrm{mod} \\; N) = 1$$ and $k$ is a random integer between $0$ and $r - 1$. From this equation, we have $$(a^r - 1) \\; (\\textrm{mod} \\; N) = 0,$$ which means $N$ must divide $a^r-1$. If $r$ is also even, then we can write $$a^r -1 = (a^{r/2}-1)(a^{r/2}+1).$$ If $r$ is not even, we cannot go further and must try again with a different value for $a$; otherwise, there is a high probability that the greatest common divisor of $N$ and either $a^{r/2}-1$, or $a^{r/2}+1$ is a proper factor of $N$.\n",
```